Sprint 9

superaydumov opened the pull request:
Pull request according to sprint 9 in Yandex Practicum course.

superaydumov committed:
Refactored ImageListViewController and ImageListCell classes.

superaydumov committed:
Added ProfileViewController and SingleImageViewController.
Added view controllers in storyboard.
Refactored ImagesListViewController class (added method to show single image through Navigation view controller).

superaydumov committed:
Added ability to zoom and share selected photo.
Refactored ProfileViewController (all UI elements added through code).

andrew-kuzn-yndx commented on ImageFeed/ImagesList/ImageListViewController.swift:
```diff
@@ -38,36 +40,16 @@ class ImagesListViewController: UIViewController {
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?){
+        if segue.identifier == showSingleImageSegueIdentifier {
+            let viewController = segue.destination as! SingleImageViewController
+            let indexPath = sender as! IndexPath
```
В этих строках есть force операция - force cast, она как и force unwrap является нежелательной и надо стараться такого кода избегать. Можно переписать без их использования, переместив в мягкий каст в if: if let a = a as? 
Всегда когда есть возможность написать безопасную конструкцию – отдавай предпочтение ей. Так твоя работа будет выигрышнее смотреться)

andrew-kuzn-yndx commented on ImageFeed/ImagesList/ImageListViewController.swift:
```diff
@@ -79,22 +61,26 @@ extension ImagesListViewController: UITableViewDataSource {
+        let date = dateFormatter.string(from: Date())
+        let isLiked = indexPath.row % 2 == 0
+        
+        cell.configureCell(image: image, date: date, isLiked: isLiked)
```
Молодец, что перенес этот метод внутрь ячейки :)

andrew-kuzn-yndx commented on ImageFeed/ImagesList/ImagesListCell.swift:
```diff
@@ -10,9 +10,35 @@ import UIKit
+        gradientView.backgroundColor = UIColor.clear
+        gradientView.layer.insertSublayer(gradientViewLayer, at: 0)
+        gradientView.layer.maskedCorners = [.layerMinXMaxYCorner, .layerMaxXMaxYCorner]
+    }
```
А вот такие штуки, как методы для создания градиента - лучше выводить в отдельный экстенш. Ведь теоретически создание градиента может пригодиться в разных частях приложения

andrew-kuzn-yndx commented on ImageFeed/ImagesList/ProfileViewController.swift:
```diff
@@ -0,0 +1,95 @@
+        logoutButtonUISetup()
+        nameLabelUISetup()
+        loginLabelUISetup()
+        infoLabelUISetup()
```
Это немного устаревший и сложный метод создания вью. Все свойства вью можно задавать при инициализации, например:

private lazy var nameLabelUISetup: UILabel {
       let nameLabel = UILabel()
         nameLabel.text = "Name label"
         nameLabel.font = UIFont.systemFont(ofSize: 23, weight: .semibold)
         nameLabel.textColor = .ypWhite
         nameLabel.translatesAutoresizingMaskIntoConstraints = false
         return nameLabel
}()

superaydumov replied:
 Попробовал реализовать, но при создании функции с констреинтами ругается на то, что не видит свойство nameLabel. Как это можно исправить?

andrew-kuzn-yndx commented on ImageFeed/ImagesList/ProfileViewController.swift:
```diff
@@ -0,0 +1,95 @@
+            imageView.heightAnchor.constraint(equalToConstant: 70),
+            imageView.widthAnchor.constraint(equalToConstant: 70),
+            imageView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 32),
+            imageView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: 16)])
```
Добавлять сабвью и констрейнты лучше в отдельном методе, где они буду вместе в одном месте)

andrew-kuzn-yndx commented on ImageFeed/ImagesList/SingleImageViewController.swift:
```diff
@@ -0,0 +1,72 @@
+        }
+    }
+    // MARK: - IBOutlets
+    @IBOutlet weak var singleImageView: UIImageView!
```
Можно указать private

andrew-kuzn-yndx commented on ImageFeed/ImagesList/SingleImageViewController.swift:
```diff
@@ -0,0 +1,72 @@
+        dismiss(animated: true, completion: nil)
+    }
+    
+    @IBAction func didTapShareButton(_ sender: Any) {
```
обязательно нужно private

superaydumov committed:
Created additional extension in case gradient method is needed somewhere else.
Added private property to outlets and action in SingleImageViewController.
Replaced cell selection style in storyboard (default to none).

superaydumov committed:
Refactored "prepare" method in ImageListViewController (removed force unwrapping).
